openwisp_monitoring/monitoring/api/views.py

```diff
@@ -0,0 +1,183 @@
+                    continue
+                chart_dict['description'] = chart.description
+                chart_dict['title'] = chart.title.format(
+                    metric=chart.metric, **chart.metric.tags
+                )
+                chart_dict['type'] = chart.type
+                chart_dict['unit'] = chart.unit
+                chart_dict['summary_labels'] = chart.summary_labels
+                chart_dict['colors'] = chart.colors
+                chart_dict['colorscale'] = chart.colorscale
+                for attr in ['fill', 'xaxis', 'yaxis']:
+                    value = getattr(chart, attr)
+                    if value:
+                        chart_dict[attr] = value
+            except InvalidChartConfigException:
+                logger.exception(f'Skipped chart for metric {chart.metric}')
+                continue
+            # get x axys (only once)
+            if x_axys and chart_dict['x'] and chart.type != 'histogram':
+                data['x'] = chart_dict.pop('x')
+                x_axys = False
+            # prepare to sort the items according to
+            # the order in the chart configuration
+            key = f'{chart.order} {chart_dict["title"]}'
+            chart_map[key] = chart_dict
+        # add sorted chart list to chart data
+        data['charts'] = list(OrderedDict(sorted(chart_map.items())).values())
+        return data
+
+    def _get_csv(self, data):
+        header = ['time']
+        columns = [data.get('x')]
+        histograms = []
+        for chart in data['charts']:
+            if chart['type'] == 'histogram':
+                histograms.append(chart)
+                continue
+            for trace in chart['traces']:
+                header.append(self._get_csv_header(chart, trace))
+                columns.append(trace[1])
+        rows = [header]
+        for index, element in enumerate(data.get('x', [])):
+            row = []
+            for column in columns:
+                row.append(column[index])
+            rows.append(row)
+        for chart in histograms:
+            rows.append([])
+            rows.append([chart['title']])
+            # Export value as 0 if it is None
+            for key, value in chart['summary'].items():
+                if chart['summary'][key] is None:
+                    chart['summary'][key] = 0
+            # Sort Histogram on the basis of value in the descending order
+            sorted_charts = sorted(
+                chart['summary'].items(), key=lambda x: x[1], reverse=True
+            )
+            for field, value in sorted_charts:
+                rows.append([field, value])
+        # write CSV to in-memory file object
+        fileobj = StringIO()
+        csv.writer(fileobj).writerows(rows)
+        return fileobj.getvalue()
+
+    def _get_csv_header(self, chart, trace):
+        header = trace[0]
+        return f'{header} - {chart["title"]}'
+
+
+dashboard_timeseries = DashboardTimeseriesView.as_view()
```

Comment:
I think as last step we shall try to remove the duplication between this view and the device monitoring view

Reply:
Yes, I have already planned for this. https://github.com/openwisp/openwisp-monitoring/pull/376#discussion_r878154238